### PR TITLE
Lazy initialize api to initialize target types

### LINKED ISSUE #8738: Use LazyIntializer API to initialize a target type in a multi threaded environment
## Details about Problem
In the current [implementation](https://github.com/NuGet/NuGet.Client/blob/c1f6915918b82c096bbd666bd9c18528f1f70630/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/ExtensibleSourceRepositoryProvider.cs#L54) object is always locked even-though target type is already initialized.  Learnt from this [doc](https://github.com/microsoft/referencesource/blob/17b97365645da62cf8a49444d979f94a59bbb155/mscorlib/system/threading/LazyInitializer.cs#L197) that there is a .NET framework implementation to accomplish the same logic. `This issue is created to track other lazy initializations which rely on a similar pattern`.

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): VS UI

### PR DESCRIPTION
**Pull request by erdembayar**
## Bug

Fixes: [NuGet/Home/#8738](https://github.com/NuGet/Home/issues/8738)
Regression: No

## Fix
Details: object is always locked even-though PackageSourceProvider instance is already initialized in the current implementation. Learnt from this [doc ](https://github.com/microsoft/referencesource/blob/17b97365645da62cf8a49444d979f94a59bbb155/mscorlib/system/threading/LazyInitializer.cs#L197) that there is a .NET framework implementation to accomplish the same logic. Hence replaced custom locking mechanism code with framework API.

## Testing/Validation
Tests Added: No
Reason for not adding tests: Tests are not added because of the nature of the fix.
Validation: Telemetry data has to be analyzed to test the accuracy of this fix once the code ships in the upcoming preview release.

### CODE REVIEW

**Review by donnie-msft — APPROVED** [review]
Seems comparable to Kartheek's work in this area 👍

**Comment by zivkan** [comment]
Neither the issue nor the PR explains what the benefit of using this API is.

It's very subjective, but I find the manual locking easier to understand because I can see what the code code, whereas I need to look up API docs to see what all the arguments mean.

Some of the changed code wraps anonymous functions that call JTF and need the UI thread. Since `LazyInitializer.EnsureInitialized` is in the BCL, it doesn't know anything about JTF.  Is there potential for problems?  When `LazyInitializer.EnsureInitialized` needs to run the value factory, what thread does it run it on?

**Comment by kartheekp-ms** [comment]
I think the `issue` fixed and `testing/validation` in this PR needs to be updated because the code paths fixed were different.

> It's very subjective, but I find the manual locking easier to understand because I can see what the code code, whereas I need to look up API docs to see what all the arguments mean.

I agree. It looks like in the current implementation `object is locked` even though the `target object` has already been initialized. `LazyInitializer.EnsureInitialized` locks the object only when the `target` type is not initialized saving couple of milliseconds. I think using the .NET API produces cleaner code.

> Some of the changed code wraps anonymous functions that call JTF and need the UI thread. Since LazyInitializer.EnsureInitialized is in the BCL, it doesn't know anything about JTF. Is there potential for problems? When LazyInitializer.EnsureInitialized needs to run the value factory, what thread does it run it on?

Good question. As per the [implementation](https://github.com/microsoft/referencesource/blob/master/mscorlib/system/threading/LazyInitializer.cs#L241), it looks like `valueFactory` is invoked on the `caller` thread. Since `valueFactory` is a delegate of type `Func<T>`, I guess it should be okay to use JTF API in the anonymous method unless @erdembayar noticed different behavior when manually validating this fix.

**Comment by srdjanjovcic** [comment]
Well, it seems that EnsureInitialized is simply a wrapper around double-lock pattern, which itself is also replaced with Lazy<T>. Benefits of LazyInitializer are that there is no separate object (of Lazy<T>) which may be beneficial if object holding Lazy<T> is itself frequently created. For JTF concerns, there is AsyncLazy<T>, which accepts JTF as a parameter to be used during initialization.

---
In reply to: [646874202](https://github.com/NuGet/NuGet.Client/pull/3447#issuecomment-646874202) [](ancestors = 646874202)

**Review by srdjanjovcic — APPROVED** [review]
:shipit:

File: src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs:207]: One of the remarks in https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized?view=netcore-3.1, scares me. 

In particular: 

```
In the event that multiple threads access this method concurrently, multiple instances of T may be created, but only one will be stored into target. In such an occurrence, this method will not dispose of the objects that were not stored. If such objects must be disposed, it is up to the caller to determine if an object was not used and to then dispose of the object appropriately.

```

I don't know if the initialization below can handle that. 

Now the likelihood of concurrent initialization like is super low, but it does give me a pause.

File: src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs:162]: Same as the other one.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs
Comment (srdjanjovcic) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs:207]: Maybe `Lazy<T>` is better choice here, as it is guaranteed to be thread-safe, and single-initialized.

> By default, `Lazy<T>` objects are thread-safe. [...] In multi-threaded scenarios, the first thread to access the Value property of a thread-safe `Lazy<T>` object initializes it for all subsequent accesses on all threads, and all threads share the same data. 

https://docs.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs:207]: A lazy would deadlock if called from a background thread. Think we already have that problem with extensiblesourceprovider. 

Maybe the answer is to not do anything here.

Comment (kartheekp-ms) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs:207]: > One of the remarks in https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized?view=netcore-3.1, scares me.
> 
> In particular:
> 
> ```
> In the event that multiple threads access this method concurrently, multiple instances of T may be created, but only one will be stored into target. In such an occurrence, this method will not dispose of the objects that were not stored. If such objects must be disposed, it is up to the caller to determine if an object was not used and to then dispose of the object appropriately.
> ```
> 
> I don't know if the initialization below can handle that.

As per my understanding, the [`overloaded method of EnsureInitialized`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized?view=netcore-3.1#System_Threading_LazyInitializer_EnsureInitialized__1___0__System_Boolean__System_Object__System_Func___0__) invoked here is thread safe and multiple instances of `T` will **not** be created if the method is accessed concurrently.

Comment (srdjanjovcic) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs:207]: @nkolev92 - I didn't know that `Lazy<T>` would deadlock... Did you encounter it yourself? I am asking because we are extensively use for singletons in VS code base 😐

Maybe `AsyncLazy<T>` would be better? You can also pass JTF to it...

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs:207]: If a lazy has any thread affinity it can easily lead to a deadlock.
Ex: The lazy itself is called on the UI thread, but the code inside the tries to switch to background/ui thread at some point. See https://github.com/NuGet/NuGet.Client/pull/2919. 

I'll try to come back to this tomorrow.

Comment (srdjanjovcic) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs:207]: Thanks, @nkolev92 -- yeah, it seems that `AsyncLazy<T>` is preferred over `Lazy<T>`

Comment (erdembayar) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs:207]: @nkolev92 there is no `AsyncLazy` version of this `LazyInitializer.EnsureInitialized`, also I don't see we really need this change.  Do we just abandon this PR?

File: src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs:162]: @nkolev92 there is no `AsyncLazy` version of this `LazyInitializer.EnsureInitialized`, also I don't see we really need this change.  Do we just abandon this PR?

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs:162]: I need to do some tests but I haven't had the time to dig in. I'll try to do that soon so I unblock this PR. 
Sorry for the delays, there are a bunch of higher pri PRs that I had to look at (ones affecting 5.7)

Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs:162]: @nkolev92 We need to make decision on this PR. It has been here for 2 months. Maybe just abandon it since there is no actual need or urgency of implement this change. I feel it's nice to have thing.

Comment (zivkan) [reply src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs:162]: > there is no AsyncLazy version of this LazyInitializer.EnsureInitialized

Consider the semantic meaning of `LazyInitializer.EnsureInitialized` and how that behavior differs from `AsyncLazy`. Aren't they the same?

Comment (kartheekp-ms) [reply src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs:162]: > also I don't see we really need this change. Do we just abandon this PR?

I support this change because it avoids locking the object every time `Start()` method is invoked even though `value` has been initialized already. As per my understanding, the [`overloaded method of EnsureInitialized`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized?view=netcore-3.1#System_Threading_LazyInitializer_EnsureInitialized__1___0__System_Boolean__System_Object__System_Func___0__) invoked here is thread safe. [LazyIntializer doc](https://docs.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization).

File: src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs:207]:
```cs
namespace Microsoft.VisualStudio.Threading
{
    //
    // Summary:
    //     Lazily executes a delegate that has some side effect (typically initializing
    //     something) such that the delegate runs at most once.
    public class AsyncLazyInitializer

```

Visual Studio threading has an AsyncLazyInitializer. 

LazyInitializer will deadlock when the blocked thread is a UI thread and the initialization inside needs the UI thread.

Honestly, while this was a great exercise, I'm not sure there'll be an actual benefit. Don't fix what's not broken.

File: src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Console/Console/ConsoleDispatcher.cs:162]: Left a comment on the other thread as well, when I posted the original comment I was looking at the wrong description.

My point about a potential deadlock still stands. Lazy initializer won't play well when the UI thread is blocked on the initialization, yet internally the initalization step needs the UI thread.

**Comment by erdembayar** [comment]
@kartheekp-ms  @nkolev92 
Since issue is closed I am closing PR too, it might introduce new deadlock issue.
https://github.com/NuGet/Home/issues/8738#issuecomment-712483416

**Pull request closed without merging**
